### PR TITLE
Fix tools not being deduplicated | Fix bug where tools logging causes nil pointer exception

### PR DESCRIPTION
**Pull request by taimoorgit**
## Issues

https://github.com/OpsLevel/kubectl-opslevel/issues/210

https://github.com/OpsLevel/opslevel-jq-parser/pull/14 and https://github.com/OpsLevel/opslevel-jq-parser/pull/15

## Changelog

- [x] Add a unit test that covers all cases
- [x] Make a `changie` entry
- [x] Added a a nil check for `*tool.Environment`. A customer reported this was crashing for them.

### CODE REVIEW

File: src/common/reconciler.go [hunk 1]
```diff
@@ -293,7 +293,23 @@ func (r *ServiceReconciler) handleCreateTags(service *opslevel.Service, registra
 }
 
 func (r *ServiceReconciler) handleTools(service *opslevel.Service, registration opslevel_jq_parser.ServiceRegistration) {
+	set := map[string]struct{}{} // necessary for deduplication of tools created
```
Comment (rocktavious) [inline src/common/reconciler.go:296]:
```suggestion
	set := make(map[string]struct{}) // necessary for deduplication of tools created
```

File: src/common/reconciler.go [hunk 2]
```diff
@@ -293,7 +293,23 @@ func (r *ServiceReconciler) handleCreateTags(service *opslevel.Service, registra
...
+		if tool.Environment != nil {
+			toolEnv = *tool.Environment
+		}
+		key = string(tool.Category) + string(tool.DisplayName) + string(toolEnv)
+		if _, ok = set[key]; ok {
+			log.Debug().Msgf("[%s] Tool '{Category: %s, Environment: %s, Name: %s}' this is a duplicate entry ... skipping", service.Name, tool.Category, *tool.Environment, tool.DisplayName)
+			continue
+		}
+		set[key] = struct{}{}
+
```
[the remark this answers is not in the document]
Comment (taimoorgit) [reply src/common/reconciler.go]: @rocktavious I added this change. Here's where the code now lives in `opslevel-jq-parser`: https://github.com/OpsLevel/opslevel-jq-parser/pull/14/files#diff-a16bc4e19331b5487156b26be9c572a4542f2e79ce787d2c2504cbfe7b46e4b5

File: src/submodules/opslevel-common
Comment (davidbloss) [inline src/submodules/opslevel-common:1]: Archived repo

File: src/common/reconciler_test.go [hunk 3]
```diff
@@ -403,3 +403,61 @@ func Test_Reconciler_ServiceNeedsUpdate(t *testing.T) {
 		autopilot.Equals(t, test.result, reconciler.ServiceNeedsUpdate(test.input, &test.service))
 	})
 }
+
+func Test_Reconciler_HandleTools(t *testing.T) {
+	// Arrange
+	var (
```
Comment (davidbloss) [inline src/common/reconciler_test.go:409]: Can we please pull most of this out of `var ()`?

Comment (taimoorgit) [reply src/common/reconciler_test.go:409]: @davidbloss will do

**Review by rocktavious — APPROVED** [review]

**Pull request merged**